### ISSUE TITLE
Clarification of weights for animations

### ISSUE THREAD

**Issue by Fabrice3D** [KhronosGroup/glTF#1028]
so I got my parser and exporter working for TRS keyframes
also works nicely on Sketchfab (as far as nice this anim can be)
https://sketchfab.com/models/7f47f3fc84174a8bbe0e375f6a21e6a5

Next was to implement the new 2.0 morph. Unfortunatly, I do not get the mechanics of the weights from the provided examples.
For morphing/vertexAnimations, my program uses mechanics similar to md2 format. Where 2 sources vertices (or more are being used) and interpolated based on duration/ elapsed time . Looking at the stored weight 0-1 values, I can extrapolate its the interpolation value itself. But I'm lost at what and where the sources are.
Also I can't place why the provided example AnimatedMorphCube holds so many keyframes, like 30 per second. Looking at the provided example, 4 or 5 should be enough to reproduce the gif. So It's clearly intended to work in a way I do not grasp.

Aside this page explaining the storage, I wasn't able to find any info on this.
https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#animations

So if someone could explain me in big lines how its supposed to work.
That would be very helpful. thx by advance.

**Comment by lexaknyazev** [comment]
@Fabrice3D 
Each element of `mesh.primitive.targets` array contains per-vertex differences from the base mesh data. 
Were you able to load targets w/o animations?

**Comment by Fabrice3D** [comment]
![screen shot 2017-06-28 at 14 17 02](https://user-images.githubusercontent.com/550453/27636591-8f130c78-5c0c-11e7-807a-8e34c3a1a9a6.jpg)

I see the cube, I have all the data, keyframes data, I simply do not know yet what to do with it.

**Comment by lexaknyazev** [comment]
Animation output accessor has following layout:
```
n - number of weights
c - number of keyframes

[W0, W1, ..., Wn] * c
```

To get morphed mesh, implementation should add weighted sum of all targets to the base mesh.

**Comment by Fabrice3D** [comment]
aaaah! Starts makes sense now. Thank you Lexaknyazev.

**Comment by javagl** [comment]
No formulas, but a snippet from the upcoming glTF Overview for glTF 2.0:

![gltfmorphtargets](https://user-images.githubusercontent.com/5597569/27641394-f84c0b90-5c1b-11e7-876e-32adea9a526f.png)

(Hoping that this appropriately conveys the basic, intuitive idea...(?))

**Comment by Fabrice3D** [comment]
Thank you javagl, helps a lot.
"each primitives can contain an array", I'm still in darkness here on this one.
I mean, I now get the concept thanks to both replies and I should not need to burn an excessive amount of neurones to make it work when it uses just two sources. But how that goes if you have more? What is the mechanism to toggle sources from keyframe? or is it limited to 2?

**Comment by javagl** [comment]
I'm not sure what you mean by "toggle". But an example (from my understanding - I also haven't *implemented* this yet...) 

    Rendered POSITIONS = 
        mesh primitive POSITIONS +
          weights[0] * morph target 0 POSITIONS +
          weights[1] * morph target 1 POSITIONS +
          ...
          weights[N] * morph target N POSITIONS +

    For example:

    mesh primitive POSITIONS: [1.0, 2.0, 3.0]
    morph target 0 POSITIONS: [2.0, 0.0, 0.0]
    morph target 1 POSITIONS: [0.0, 3.0, 0.0]
    morph target 2 POSITIONS: [0.0, 0.0, 4.0]

    Weights = (0.25, 1.0, 0.5) 
    
    Renderered POSITIONS =
               [1.0, 2.0, 3.0] +
        0.25 * [2.0, 0.0, 0.0] +
        1.0  * [0.0, 3.0, 0.0] +
        0.5  * [0.0, 0.0, 4.0]
      =        [1.5, 5.0, 5.0) 
        
So basically:

- Compute a linear combination of the morph target positions (where the factors are given by the weight elements)
- Add this to the original positions

I hope that I'm not overlooking something here.

**Comment by lexaknyazev** [comment]
> What is the mechanism to toggle sources from keyframe? or is it limited to 2?

Each keyframe contains weights for all targets (in the same order as `mesh.primitive.targets`). 
Depending on hardware/software caps, implementations could use different strategies to "toggle" targets, such as: always apply all (incl. targets with zero weight), filter out targets with weights below certain margin, etc...

**Comment by Fabrice3D** [comment]
![weights](https://user-images.githubusercontent.com/550453/27645076-85f2fb12-5c25-11e7-8a8e-c8431b46b80b.jpg)

ah! just see your replies.
was meanwhile busy digesting the concept...  seems its equivallent to javaGl explanations

The "toggle", was more thought from the editing point of view, as I will prolly offer a way to import geometries (of same order and vertices count). Not just move vertices around from a given mesh.

Thank you both, I think I should have that baby up&running in a not too far future.

**Comment by javagl** [comment]
The image might cause a misunderstanding, though. In the text, it is described properly, but the images do not match that: The morph targets do not contain the *positions*, but only the *displacements*. (That is what I tried to sketch using these arrows in my image). And this is also why the morph target data can be stored efficiently in *sparse accessors*: For the vertices where not displacement is applied, no data has to be stored.

**Comment by Fabrice3D** [comment]
ah yes, its indeed confusing but was just a quick graphic ment for me to help me link visually the relationships and make sense of previous replies and ask for confirmation. Which you did just before I posted it.

I think there is a need for an official version of this. In fact, (lots of) simple graphics would help in general sense for this format a lot, as strings id's have been replaced (for the good) with indices.

**Comment by Fabrice3D** [comment]
k, after loosing time trying access the displacements factors based on vertex indices, I finally got the sphere and the cube morph examples run after noticing the length of displacements buffer was the keyframes times buffer length * 2...
However while all looks and run fine code wise, I do not have the same animation of the cube as seen on the gif file. On mine the moving side (along z axis) of the cube goes beyond the opposite side almost double original geom depth.

The sphere seams similar to the example gif, but as its a sphere first and on extreme morph very 'artistic'. Its hard to see for sure.

My program being left handed? Is there something I should consider to transform?

To compair I've tried load with Threejs parser, but both example fails with it and Sketchfab doesn't support this feature yet. So, how could I get some compair/validation?
If there's nothing avail atm, would it be too much to ask for one small extra anim, like an extruded number or a character (not a 8 or i :)) ) to have some visual reference regarding the orientation?
+ start and end of the morph as pict?

**Comment by pjcozzi** [comment]
@Fabrice3D thanks for starting this discussion!

Discussed on the 3D Formats call.

@javagl is making a really simple morph target example for the [glTF tutorials](https://github.com/javagl/glTF-Tutorials/tree/master/gltfTutorial#gltf-tutorial) to cover displacement, etc.

**Comment by Fabrice3D** [comment]
@pjcozzi Where was it when i needed it! :) I'm sure it's gonna be informative.
Meanwhile got own generator for static and animated results working on my own format.
Not coupled yet to my glTF exporter (clients projects have prio), but should be in not too far
future...

![morphs](https://user-images.githubusercontent.com/550453/31786729-b562e136-b509-11e7-8229-0047f49af490.jpg)

**Comment by pjcozzi** [comment]
@Fabrice3D very cool, please add your work-in-progress to #1058 so it can make its way to the main glTF tools list: https://github.com/KhronosGroup/glTF/blob/master/README.md#gltf-tools

**Comment by zellski** [comment]
I wonder if it wouldn't be helpful to add another sentence on the expected structure of the 'weights' output channel. There is enough information in the spec to figure it out, but just barely: there isn't any left-over to confirm your assumptions. In most places, weights are represented as arrays. For other types of animations, we're outputting GL VECs. In the morph animation output stream, however, what is logically an array of arrays should instead be flattened into a single stream of scalars. I understand why this has to be so, but it's just weird enough a situation as to be unnerving without confirmation.

We already have *A Morph Target animation frame is defined by a sequence of scalars of length equal to the number of targets in the animated Morph Target.* which is fine. Perhaps simply add something like, *These scalar sequences must lie end-to-end as a single stream in the output accessor, whose final size will be the number of morph targets times the number of animation frames.*

Thoughts? Am I just dense? Is this, in fact, obvious to everyone else? :-)

**Comment by pjcozzi** [comment]
@zellski a clarification like this in the spec would be very welcome.  Feel free to open a pull request.

**Comment by javagl** [comment]
A section about morph targets has been added in the tutorial: https://github.com/javagl/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_017_SimpleMorphTarget.md 

However, the fact that the weights are always declared as `SCALAR` also confused me a bit during the implementation. A few words about this might be added either in the spec or the tutorial. (Leaving this issue open until the decision is made)

**Comment by pjcozzi** [comment]
@zellski @javagl pull requests to the spec are still very welcome! 😄

**Comment by lexaknyazev** [comment]
Closing, since the spec has been updated.
